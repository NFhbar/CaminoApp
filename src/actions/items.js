import uuid from 'uuid';

// ADD_ITEM
export const addItem = (
  {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = {}
) => ({
  type: 'ADD_ITEM',
  item: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt
  }
});

// REMOVE_ITEM
export const removeItem = ({ id } = {}) => ({
  type: 'REMOVE_ITEM',
  id
});

// EDIT_ITEM
export const editItem = (id, updates) => ({
  type: 'EDIT_ITEM',
  id,
  updates
});
