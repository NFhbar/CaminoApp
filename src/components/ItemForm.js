import React from 'react';
import { alert, button } from 'reactstrap';

export default class ItemForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.item ? props.item.description : '',
      note: props.item ? props.item.note : '',
      amount: props.item ? (props.item.amount /100).toString() : '',
      error: ''
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide and item and amount.'}));
    } else {
      this.setState(() =>({ error: ''}));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10)*100,
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error &&
          <div className=
            "alert
            alert-warning
            alert-dismissible
            fade
            show"
            role="alert">
            {this.state.error}
          </div>
        }
        <form onSubmit={this.onSubmit}>
          <div className="form-group">
            <input
              type="text"
              placeholder="Description"
              autoFocus
              value={this.state.description}
              onChange={this.onDescriptionChange}
              className="form-control"
            />
          </div>
          <div className="form-group">
            <input
              type="text"
              placeholder="Amount"
              value={this.state.amount}
              onChange={this.onAmountChange}
              className="form-control"
              />
          </div>
          <div className="form-group">
            <textarea
              placeholder="Add a note for your item (optional)"
              value={this.state.note}
              onChange={this.onNoteChange}
              className="form-control"
              >
            </textarea>
          </div>
          <button type="submit" className="btn btn-lg btn-primary">Submit Item</button>
        </form>
      </div>
    )
  }
}
