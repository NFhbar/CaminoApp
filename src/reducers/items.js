// Items Reducer

const itemsReducerDefaultState = [];

export default (state = itemsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_ITEM':
      return [
        ...state,
        action.item
      ];
    case 'REMOVE_ITEM':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_ITEM':
      return state.map((item) => {
        if(item.id === action.id){
          return {
            ...item,
            ...action.updates
          };
        } else {
          return item;
        };
      });
    default:
      return state;
  }
};
