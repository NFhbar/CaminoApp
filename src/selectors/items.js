// Get visible items
export default (items, {text, sortBy}) => {
  return items.filter((item) => {
    const textMatch = item.description.toLowerCase().includes(text.toLowerCase());

    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'amount') {
      return a.amount < b.amount ? 1: -1;
    } else if (sortBy === 'sales') {
      return a.sales < b.sales ? 1: -1;
    }
  });
};
