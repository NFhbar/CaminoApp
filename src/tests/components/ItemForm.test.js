import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import ItemForm from '../../components/ItemForm';
import items from '../fixtures/items';

test('should render ItemForm correctly', () => {
  const wrapper = shallow(<ItemForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ItemForm correctly with item data', () => {
  const wrapper = shallow(<ItemForm item={items[1]}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
  const wrapper = shallow(<ItemForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should set description on input change', () => {
  const value = 'New description';
  const wrapper = shallow(<ItemForm />);
  wrapper.find('input').at(0).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('description')).toBe(value);
});

test('should set note on textarea change', () => {
  const value = 'New note value';
  const wrapper = shallow(<ItemForm />);
  wrapper.find('textarea').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('note')).toBe(value);
});

test('should set amount if valid input', () => {
  const value = '24.50';
  const wrapper = shallow(<ItemForm />);
  wrapper.find('input').at(1).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('amount')).toBe(value);
});

test('should not set amount if invalid input', () => {
  const value = '24.5052';
  const wrapper = shallow(<ItemForm />);
  wrapper.find('input').at(1).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('amount')).toBe('');
});

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<ItemForm item={items[0]} onSubmit={onSubmitSpy}/>);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    description: items[0].description,
    amount: items[0].amount,
    note: items[0].note,
    createdAt: items[0].createdAt
  });
});

test('should set new date on date change', () => {
  const now = moment();
  const wrapper = shallow(<ItemForm />);
  wrapper.find('SingleDatePicker').prop('onDateChange')(now);
  expect(wrapper.state('createdAt')).toEqual(now);
});

test('should set calendar focus on change', () => {
  const focused = true;
  const wrapper = shallow(<ItemForm />);
  wrapper.find('SingleDatePicker').prop('onFocusChange')({ focused });
  expect(wrapper.state('calendarFocused')).toBe(focused);
});
