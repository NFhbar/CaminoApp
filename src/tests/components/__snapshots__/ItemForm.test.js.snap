// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ItemForm correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        onChange={[Function]}
        placeholder="Add a note for your item (optional)"
        value=""
      />
    </div>
    <button
      className="btn btn-lg btn-primary"
      type="submit"
    >
      Submit Item
    </button>
  </form>
</div>
`;

exports[`should render ItemForm correctly with item data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value="Shirt"
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value="10"
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        onChange={[Function]}
        placeholder="Add a note for your item (optional)"
        value=""
      />
    </div>
    <button
      className="btn btn-lg btn-primary"
      type="submit"
    >
      Submit Item
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        onChange={[Function]}
        placeholder="Add a note for your item (optional)"
        value=""
      />
    </div>
    <button
      className="btn btn-lg btn-primary"
      type="submit"
    >
      Submit Item
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 2`] = `
<div>
  <div
    className="alert alert-warning alert-dismissible fade show"
    role="alert"
  >
    Please provide and item and amount.
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        onChange={[Function]}
        placeholder="Add a note for your item (optional)"
        value=""
      />
    </div>
    <button
      className="btn btn-lg btn-primary"
      type="submit"
    >
      Submit Item
    </button>
  </form>
</div>
`;
